File: solutions/cpp/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    std::vector<std::string> letterCombinations(std::string digits) {
        if (digits.empty()) return {};

        std::string phone_map[] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        std::vector<std::string> output;
        backtrack("", digits, phone_map, output);
        return output;
    }

private:
    void backtrack(std::string combination, std::string next_digits, std::string phone_map[], std::vector<std::string>& output) {
        if (next_digits.empty()) {
            output.push_back(combination);
        } else {
            std::string letters = phone_map[next_digits[0] - '2'];
            for (char letter : letters) {
                backtrack(combination + letter, next_digits.substr(1), phone_map, output);
            }
        }
    }
};

File: solutions/csharp/0017-letter-combinations-of-a-phone-number.cs
public class Solution {
    public IList<string> LetterCombinations(string digits) {
        if (string.IsNullOrEmpty(digits)) return new List<string>();

        string[] phone_map = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        List<string> output = new List<string>();
        Backtrack("", digits, phone_map, output);
        return output;
    }

    private void Backtrack(string combination, string next_digits, string[] phone_map, List<string> output) {
        if (next_digits.Length == 0) {
            output.Add(combination);
        } else {
            string letters = phone_map[next_digits[0] - '2'];
            foreach (char letter in letters) {
                Backtrack(combination + letter, next_digits.Substring(1), phone_map, output);
            }
        }
    }
}
